### PR TITLE
fix: Initial commit targetting grpc registry server

### LINKED ISSUE #4392: Remote registry client does not map application errors
## Expected Behavior 
A feast application has a registry of type remote connected to a registry server.
The feast client invokes a an API on the feast registry that thrown an error, e.g. `store.get_feature_service("foo")` which raises
`FeatureServiceNotFoundException` from the registry server because `foo` is not a registered feature service.

The expectation is that the method invoked by the client raises the `FeatureServiceNotFoundException` and the application can detect it with:
```py
try:
    fs = store.get_feature_service("foo")
except FeatureServiceNotFoundException
    ....
````

## Current Behavior
The registy server always raises an `_InactiveRpcError` error instead.

This bug causes unexpected errors when the client invokes services whose implementation needs such expections to execute fall-back logic, like [_get_feature_views_to_materialize](https://github.com/feast-dev/feast/blob/125fa49fc700a8e6f1994c710a5076210fb97bf9/sdk/python/feast/feature_store.py#L684).

## Steps to reproduce
See `Expected Behavior` section.

### Specifications

- Version: 0.40.0
- Platform: any
- Subsystem: registry server

## Possible Solution
The registry server can wrap the original exception in an error status message that the remote registry client can catch and translate into the original exception:
[](https://github.com/grpc/grpc/tree/master/examples/python/errors)

### PR DESCRIPTION
**Pull request by dmartinol**
# What this PR does / why we need it:
Intercept registry server failures and adds an additional message to the response with the failure status and details.
The proxy client then rebuilds the original error from the status data.

# Which issue(s) this PR fixes:
Partially fixes #4392 for Registry Server

# Misc
Need to add some UT/IT to prove the effectivness

### CODE REVIEW

File: sdk/python/feast/permissions/server/grpc.py [hunk 1]
```diff
@@ -31,7 +32,7 @@ def grpc_interceptors(
     if auth_type == AuthManagerType.NONE:
         return None
 
-    return [AuthInterceptor()]
+    return [AuthInterceptor(), ErrorInterceptor()]
```
Comment (lokeshrangineni) [inline sdk/python/feast/permissions/server/grpc.py:35]: Will it handle the exception happens in the `AuthInterceptor`?

Comment (dmartinol) [reply sdk/python/feast/permissions/server/grpc.py:35]: It won't. Specific code is needed in the `AuthInterceptor` (and maybe also use a Feast-specific error instead)

Comment (tmihalac) [reply sdk/python/feast/permissions/server/grpc.py:35]: I think that `ErrorInterceptor` should be called `if auth_type == AuthManagerType.NONE` as well

File: sdk/python/feast/permissions/server/grpc.py [hunk 2]
```diff
@@ -52,3 +53,55 @@ def intercept_service(self, continuation, handler_call_details):
...
+                request_deserializer=handler.request_deserializer,
+                response_serializer=handler.response_serializer,
+            )
+        return handler
+
+
+def _error_to_status_code(error: Exception) -> grpc.StatusCode:
+    if isinstance(error, FeastObjectNotFoundException):
+        return grpc.StatusCode.NOT_FOUND
+    if isinstance(error, FeastObjectNotFoundException):
```
Comment (tmihalac) [inline sdk/python/feast/permissions/server/grpc.py:105]: I think you meant `PermissionNotFoundException`

File: sdk/python/feast/permissions/server/grpc.py (hunk 1 above)
Comment (dmartinol) [reply sdk/python/feast/permissions/server/grpc.py:35]: @tmihalac yes, and also we need to move this interceptor in a different package, as it has nothing to do with the `feast.permission` package.
we'll review this later

File: sdk/python/feast/permissions/server/grpc.py (hunk 2 above)
Comment (dmartinol) [reply sdk/python/feast/permissions/server/grpc.py:105]: Resolved by https://github.com/feast-dev/feast/pull/4465

File: sdk/python/feast/permissions/server/grpc.py (hunk 1 above)
Comment (dmartinol) [reply sdk/python/feast/permissions/server/grpc.py:35]: > Will it handle the exception happens in the AuthInterceptor?

@lokeshrangineni due to the latest refactory, we need to implement our own FeastAuthenticationError, so I would create a separate ticket for that.

**Comment by dmartinol** [comment]
@tokoko @tmihalac @lokeshrangineni ready for review now

**Comment by tokoko** [comment]
@dmartinol I think you need to rebase

File: sdk/python/feast/registry_server.py [hunk 3]
```diff
@@ -668,3 +670,21 @@ def start_server(store: FeatureStore, port: int, wait_for_termination: bool = Tr
         server.wait_for_termination()
     else:
         return server
+
+
+def grpc_interceptors(
+    auth_type: AuthManagerType,
+) -> Optional[list[grpc.ServerInterceptor]]:
+    """
+    A list of the authorization interceptors.
```
Comment (tokoko) [inline sdk/python/feast/registry_server.py:679]: remove "authorization" from the line

File: sdk/python/feast/permissions/server/grpc.py (hunk 1 above)
Comment (dmartinol) [reply sdk/python/feast/permissions/server/grpc.py]: @lokeshrangineni created https://github.com/feast-dev/feast/issues/4466

**Comment by tokoko** [comment]
I'll force a retry, but that looks like a genuine credentials error

File: sdk/python/feast/errors.py [hunk 4]
```diff
@@ -1,21 +1,71 @@
...
+        m = {
+            "module": f"{type(self).__module__}",
+            "class": f"{type(self).__name__}",
+            "message": f"{str(self)}",
+        }
+        return json.dumps(m)
+
+    @staticmethod
+    def from_error_detail(detail: str) -> Optional["FeastError"]:
+        import importlib
```
Comment (tokoko) [inline sdk/python/feast/errors.py:50]: These are standard lib libraries, why not import top-level?

**Review by tokoko — APPROVED** [review]
lgtm

**Pull request merged**